### ISSUE TITLE
Default debug blacklist/whitelist for new projects?

### ISSUE THREAD

**Issue by kaphert** [laravel/ideas#1236]
Based on this pull request https://github.com/laravel/framework/pull/21336 and this comment https://github.com/laravel/framework/pull/21336#issuecomment-370995333 

As suggested in the pull request. Would it be a good idea if we hide default the credentials from the whoops debug page by default? 

It can prevent accidental password leaks for projects that are deployed to a live environment with debug mode still on.

**Comment by BrandonSurowiec** [comment]
Sounds like a good idea to me. Go for it.

**Comment by tontonsb** [comment]
The problem seems to be that the Whoops package doesn't currently support whitelisting...

**Comment by PieterjanDeClippel** [comment]
I just wrote a script that can do this.

WhoopsHandler.php:61
```
protected function registerBlacklist($handler)
{
    $whitelist = config('app.debug_whitelist', []);
    $all_superglob  = [
        '_GET'      =>  $_GET ?? [],
        '_POST'     =>  $_POST ?? [],
        '_FILES'    =>  $_FILES ?? [],
        '_COOKIE'   =>  $_COOKIE ?? [],
        '_SESSION'  =>  $_SESSION ?? [],
        '_SERVER'   =>  $_SERVER ?? [],
        '_ENV'      =>  $_ENV ?? [],
    ];

    $missing_superglob = array_diff_key($all_superglob,$whitelist);
    foreach ($missing_superglob as $missing_key => $values) {
        // If you want to be extra sure not to overrite anything:
        if (!array_key_exists($missing_key, $whitelist)) {
            $whitelist[$missing_key] = [];
        }
    }

    foreach ($whitelist as $superglob_key => $whitelisted_items) {
        $blacklist = array_except($all_superglob[$superglob_key], $whitelisted_items);
        foreach ($blacklist as $secret => $value) {
            $handler->blacklist($superglob_key, $secret);
        }
    }

    return $this;
}
```

**Comment by PieterjanDeClippel** [comment]
https://github.com/laravel/framework/pull/27552/commits/65e6d16d5c3482088e0df960df313e7a54a4e431

**Comment by tontonsb** [comment]
The problem with packages is that the stuff can get dumped before packages are registered. I did a [similar package](https://github.com/glaivepro/hidevara), but I couldn't find a reasonable way to register it soon enough. Currently I have it so that the developer must call my class in `bootstrap/app.php`.

The only "automatic" way I could think of was to edit the bootstrap automatically. Maybe I'll do that later.

**Comment by PieterjanDeClippel** [comment]
Couldn't we creatie a package that determines if there's a debug_whitelist config, and then sets the debug_blacklist config? Though this would mean that this package must be loaded before the filp/whoops package... which is yet another thing

**Comment by kaphert** [comment]
Note for anyone reading this. A simple approach is to just add this to you `config/app.php` file: 

```
'debug_blacklist' => [
        '_COOKIE' => array_keys($_COOKIE),
        '_SERVER' => array_keys($_SERVER),
        '_ENV' => array_keys($_ENV),
    ],
```

This will hide the most sensitive data from your whoops error page. My suggestion in this ticket was to make something like this the default setting.

**Comment by VishwasShetty** [comment]
> Note for anyone reading this. A simple approach is to just add this to you `config/app.php` file:
> 
> ```
> 'debug_blacklist' => [
>         '_COOKIE' => array_keys($_COOKIE),
>         '_SERVER' => array_keys($_SERVER),
>         '_ENV' => array_keys($_ENV),
>     ],
> ```
> 
> This will hide the most sensitive data from your whoops error page. My suggestion in this ticket was to make something like this the default setting.

Sir actually i tried your code in my project.but it didnt worked.
My frame work is 5.8.2.
can you help mw out.

**Comment by BrandonSurowiec** [comment]
@VishwasShetty install Laravel Ignition to be your error page and you don't need to worry about the blacklist anymore: https://flareapp.io/docs/ignition-for-laravel/installation

**Comment by tontonsb** [comment]
That is wrong. Sometimes the app crashes before Ignition is started but after Whoops is started, so one can still get their variables dumped in a most unexpected way.

**Comment by Chris1234567899** [comment]
This should be definitely hidden by default, no matter what!
Maybe people are aware of setting `debug` for production, but that does not apply equally in a test or development environment where debugging might be needed. Sensitive information (shared keys for external services for example) can be stored there as well.
I myself was not aware, that .env variables are shown by default if debug is set to true (even the thought seemed to ridiculous to be true). Huge security flaw!

**Just stating that somewhere in the docs is simply not good enough**. People might not read docs entirely or forget about it.

**Comment by JohnnyWalkerDesign** [comment]
As previous stated, you can add the following to `config/app.php`:

```
    'debug_blacklist' => [
        '_COOKIE' => array_keys($_COOKIE),
        '_SERVER' => array_keys($_SERVER),
        '_ENV' => array_keys($_ENV),
    ],
```

This DEFINITELY works.

And with Laravel 7 and 8 they added `debug_hide`, which does exactly the same thing.
